From 8236b2cf91ee5dc1bcd206c08b8e813b77f6003d Mon Sep 17 00:00:00 2001
From: rhb2 <robert.bogart@joyent.com>
Date: Wed, 11 Sep 2019 19:06:32 +0000
Subject: [PATCH] MANTA-4542 Implement alternate process_task() functions for
 rebalancer agent Reviewed by: Rui Loura <rui.loura@joyent.com> Approved by:
 Rui Loura <rui.loura@joyent.com>

---
 Cargo.toml   |  8 ++++++++
 README.md    | 18 ++++++++++++++++++
 src/agent.rs |  9 ++++++++-
 3 files changed, 34 insertions(+), 1 deletion(-)

diff --git a/Cargo.toml b/Cargo.toml
index aff4b40..97d02cc 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -7,6 +7,14 @@ edition = "2018"
 [profile.dev]
 panic = "abort"
 
+[features]
+# No features by default
+default = []
+
+# Always mark a task as complete without attempting to even download the object
+# within it.  This is for TESTING PURPOSES ONLY!
+always_pass = []
+
 [dependencies]
 base64 = "0.10.1"
 uuid = { version = "0.7.4", features = ["v4"] }
diff --git a/README.md b/README.md
index 5a77998..8b16b52 100644
--- a/README.md
+++ b/README.md
@@ -66,3 +66,21 @@ cargo check
 cargo clippy
 cargo test
 ```
+
+## Testing
+
+There is a certain flavor of the rebalancer agent that allow for more convenient
+testing of the zone -- namely one that receives a (properly formed) assignment
+and blindly marks all tasks within it as "Complete" instead of actually doing
+the leg work of processing each task.  This is intended for scenarios where
+functional verification of the happy path in the rebalancer zone.  As this
+project evolves, other modes will likely be introduced.
+
+To build the version of the agent described above, a special flag must be
+passed to the compiled to enable the feature:
+
+```
+cargo build --features "always_pass"
+```
+
+Note: By default, this feature will never be enabled.
diff --git a/src/agent.rs b/src/agent.rs
index 8f21e9b..24f4a35 100644
--- a/src/agent.rs
+++ b/src/agent.rs
@@ -620,6 +620,13 @@ fn download(
     }
 }
 
+#[cfg(feature = "always_pass")]
+fn process_task(task: &mut Task) {
+    task.set_status(TaskStatus::Complete);
+    return;
+}
+
+#[cfg(not(feature = "always_pass"))]
 fn process_task(task: &mut Task) {
     let file_path = manta_file_path(&task.owner, &task.object_id);
     let path = Path::new(&file_path);
@@ -953,7 +960,7 @@ mod tests {
             match opt {
                 None => {
                     if expected != TaskStatus::Complete {
-                        panic!("Missing expected list of failed tasks.");
+                        panic!("Assignment succeeded when it should not.");
                     }
                 }
                 Some(tasks) => {
-- 
2.21.0

